odoo/osv/expression.py

```diff
@@ -971,7 +971,7 @@ def push_result(sql):
 
             if is_operator(leaf):
                 if leaf == NOT_OPERATOR:
-                    push_result(SQL("(NOT (%s))", pop_result()))
+                    push_result(SQL("((%s) IS NOT TRUE)", pop_result()))
```

Comment:
Is it relevant? You shouldn't get a SQL leaf coming for bool anyway ?  
It affects other leaf than the bool ones for no reason (and potentially breaks custom indexes).

Reply:
Yes it is relevant. The leaf is always a SQL predicate.
Its result can be null. `NOT (some_column)` is always falsy if some_column is null, which is incorrect for the ORM.

There are no other leaves than boolean conditions (otherwise it is a syntax error). This happens only for operators that are not inversed, all basic ones have a not version so you affect only the ones that do things like sub-queries (which are not usable in custom indexes).

Reply:
> Yes it is relevant. The leaf is always a SQL predicate.

Yes, but the SQL predicate doesn't contains predicated that involved a boolean, because operators used on boolean fields are always reversible, right ?

> Its result can be null. NOT (some_column) is always falsy if some_column is null, which is incorrect for the ORM.

Do you mean that you fix (change) the semantic here ? If yes, a separate commit is needed with a test.

> all basic ones have a not version so you affect only the ones that do things like sub-queries

You change a test because of it where no subquery is involved :thinking: 
```
                   ((
                        ("test_new_api_multi__partner"."name" LIKE %s)
                        OR ("test_new_api_multi__partner"."phone" LIKE %s)
                    )) IS NOT TRUE
``

Reply:
Although I think it makes sense, I agree with @ryv-odoo: it has nothing to do with conditions of boolean fields.  Note that the evaluation of boolean field conditions is never NULL: both `x = TRUE` and `x = FALSE OR x IS NULL` are always TRUE of FALSE.  Therefore the change of the negation operator is orthogonal to this pull request.

Reply:
Separated into another PR: https://github.com/odoo/odoo/pull/186887